homeassistant/components/mqtt/discovery.py

```diff
@@ -17,8 +17,8 @@
 _LOGGER = logging.getLogger(__name__)
 
 TOPIC_MATCHER = re.compile(
-    r'(?P<prefix_topic>\w+)/(?P<component>\w+)/(?P<object_id>[a-zA-Z0-9_-]+)'
-    '/config')
+    r'(?P<prefix_topic>\w+)/(?P<component>\w+)/(?:(?P<node_id>\w+)/)?'
+    r'(?P<object_id>[\w-]+)/config')
```

Comment:
You've changed the matching algorithm of an object id to no longer match the object id validation that Home Assistant enforces. Please revert.